src/org/exist/storage/XQueryPool.java

```diff
@@ -146,23 +147,19 @@ private void returnObject(final Source source, final CompiledXQuery xquery) {
             if (pool.size() < maxPoolSize) {
                 source.setCacheTimestamp(System.currentTimeMillis());
 
-                final Deque<CompiledXQuery> stack = pool.computeIfAbsent(source, (k) -> new ArrayDeque<>());
+                final Set<CompiledXQuery> stack = pool.computeIfAbsent(source, (k) -> new ConcurrentSkipListSet<>());
 
                 if (stack.size() < maxStackSize) {
-                    // check if the query is already in pool before adding,
-                    // may happen for modules, don't add it a second time!
-                    if(!stack.contains(xquery)) {
-                        stack.push(xquery);
-                    }
+                    stack.add(xquery);
```

Comment:
Okay but now you have introduced a different concurrency issue, which is that of duplicated unnecessary work as you have two concurrent structures which are not synchronized as a unit, this is similar to the change in consistency that you introduced by adding two AtomicLong.

Consider this scenario:

1) `Thread A` executes line 150 and creates a new `ConcurrentSkipListSet` in the `pool`.
2) `Thread A` executes line 152 and enters the conditional.
3) `Thread A` is then suspended (unscheduled).
4) `Thread B` is scheduled, and executes line 150, returning the `stack` created by `Thread A`.
5) `Thread B` then executes lines 152 and 153, adding an `xquery` to the stack.
6) `Thread A` is resumed (scheduled) and now also adds the same `xquery` to the stack which `Thread B` previously added

So one issue is that in steps (5) and (6) you are repeating work when there are concurrent threads.

You also need to be aware that calling `ConcurrentSkipListSet#size()` is both an expensive operation and can be inaccurate unless care is taken to synchronize access to it.

I think there is a simpler design possible here which is also safer:

```java
private ConcurrentMap<Source, Deque<CompiledXQuery>> pool = new ConcurrentHashMap<>();
```

```java
pool.compute(source, (k, v) -> {
   if(v == null) {
      v = new ArrayDeque<>();
      v.add(xquery);
   } else if(v.size() < maxStackSize()) {
      v.add(xquery);
   }
   return v;
});  		  
```


final Set<CompiledXQuery> stack = pool.computeIfAbsent(source, (k) -> new ConcurrentSkipListSet<>());
```